src/python/dxpy/app_builder.py

```diff
@@ -212,58 +161,199 @@ def upload_resources(src_dir, project=None, folder='/', ensure_upload=False):
+                relative_dirname = dirname[len(resources_dir):]
+                dir_stat = os.lstat(dirname)
+                if not relative_dirname.startswith('/'):
+                    relative_dirname = '/' + relative_dirname
+
+                fields = [relative_dirname, str(dir_stat.st_mode), str(int(dir_stat.st_mtime * 1000))]
+                output_sha1.update(b''.join(s.encode('utf-8') + b'\0' for s in fields))
+                
+                # add an entry in the tar file for the current directory, but
+                # do not recurse!
+                subprocess.check_call(['tar', '-C', resources_dir, '-rf', tar_fh.name, '--no-recursion', '.' + relative_dirname])
+                
+                # Canonicalize the order of subdirectories; this is the order in
+                # which they will be visited by os.walk
+                subdirs.sort()
+                
+                # check the subdirectories for symlinks.  We should throw an error
+                # if there are any links that point outside of the directory (unless
+                # --force-symlinks is given).  If a link is pointing internal to 
+                # the directory (or --force-symlinks is given), we should add it 
+                # as a file.
+                for subdir_name in subdirs:
+                    dir_path = os.path.join(dirname, subdir_name)
+
+                    # If we do have a symlink,                     
+                    if os.path.islink(dir_path):
+                        # Let's get the pointed-to path to ensure that it is
+                        # still in the directory
+                        link_target = os.readlink(dir_path)
+                        is_local=(not os.path.isabs(link_target))
+                        
+                        dir_realpath=os.path.abspath(os.path.join(dirname, link_target))
+                        # make sure that the path NEVER extends outside the resources directory!
+                        d,l = os.path.split(link_target)
+                        link_parts = []
+                        while l:
+                            link_parts.append(l)
+                            d,l = os.path.split(d)
+                        curr_path = dirname
+                        for p in reversed(link_parts):
+                            curr_path = os.path.abspath(os.path.join(curr_path, p))
+                            is_local = (is_local and curr_path.startswith(resources_dir))
+                                               
+                        # If this is a local link, add it to the list of files (case Ib)
+                        # else raise an error
+                        if force_symlinks or is_local:
+                            files.append(subdir_name)
+                        else:
+                            raise AppBuilderException("Cannot include symlinks to directories outside of the resource directory.  %r points to directory %r" % (dir_path, dir_realpath))
```

Comment:
Minor, I'd prefer that error messages and the like don't include the Python "u" specifier that it attaches to strings (see the error I pasted above for an example). I suggest something like wrapping the paths in pipes.quote (which has the additional benefit of eliding quotes when they're not needed).